File: backend/src/main/java/com/twelveweekyear/feature/quarter/support/QuarterMath.java
package com.twelveweekyear.feature.quarter.support;

import com.twelveweekyear.feature.quarter.domain.QuarterState;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

/** Calendar math for quarters: bounds, day/week-of-quarter, and state. Pure & tested. */
public final class QuarterMath {

    private QuarterMath() {
    }

    public record Bounds(LocalDate start, LocalDate end, int totalDays) {
    }

    /**
     * @param state       UPCOMING / ACTIVE / COMPLETED relative to {@code today}.
     * @param currentDay  1..totalDays while ACTIVE, otherwise null.
     * @param currentWeek 1..totalWeeks while ACTIVE, otherwise null.
     */
    public record Progress(QuarterState state, Integer currentDay, Integer currentWeek,
                           int totalDays, int totalWeeks) {
    }

    /** Calendar bounds of a quarter (Q1=Jan–Mar … Q4=Oct–Dec), leap-year aware. */
    public static Bounds bounds(int year, int quarterNumber) {
        int startMonth = (quarterNumber - 1) * 3 + 1;
        LocalDate start = LocalDate.of(year, startMonth, 1);
        LocalDate end = start.plusMonths(3).minusDays(1);
        int totalDays = (int) ChronoUnit.DAYS.between(start, end) + 1;
        return new Bounds(start, end, totalDays);
    }

    public static int totalWeeks(int totalDays) {
        return (int) Math.ceil(totalDays / 7.0);
    }

    public static Progress progress(int year, int quarterNumber, LocalDate today) {
        Bounds b = bounds(year, quarterNumber);
        int totalWeeks = totalWeeks(b.totalDays());

        if (today.isBefore(b.start())) {
            return new Progress(QuarterState.UPCOMING, null, null, b.totalDays(), totalWeeks);
        }
        if (today.isAfter(b.end())) {
            return new Progress(QuarterState.COMPLETED, null, null, b.totalDays(), totalWeeks);
        }
        long dayIndex = ChronoUnit.DAYS.between(b.start(), today); // 0-based
        int currentDay = (int) dayIndex + 1;
        int currentWeek = (int) (dayIndex / 7) + 1;
        return new Progress(QuarterState.ACTIVE, currentDay, currentWeek, b.totalDays(), totalWeeks);
    }
}

File: packages/core/src/quarter/quarter-math.ts
import { parseIsoDate, toIsoDate } from "../date";
import type { GoalStatus, QuarterState } from "./types";

/**
 * Client-side port of the backend's QuarterMath + QuarterMapper (calendar quarters Q1=Jan–Mar …
 * Q4=Oct–Dec). Lets local (guest) mode derive the exact same quarter bounds, pacing, label, and
 * goal status the server would — so numbers stay consistent when a guest later syncs to the cloud.
 * ISO yyyy-MM-dd strings compare lexicographically = chronologically.
 */

const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

/** A quarter is treated as at most 13 weeks; the last week absorbs the extra 1–2 days. */
export const MAX_WEEKS = 13;

export interface QuarterBounds {
  startIso: string;
  endIso: string;
  totalDays: number;
}

/** Calendar bounds of a quarter, leap-year aware. */
export function quarterBounds(year: number, quarterNumber: number): QuarterBounds {
  const startMonthIndex = (quarterNumber - 1) * 3; // 0-based month
  const start = new Date(year, startMonthIndex, 1);
  const end = new Date(year, startMonthIndex + 3, 0); // day 0 of the next month = last day of the 3rd month
  const totalDays = Math.round((end.getTime() - start.getTime()) / 86_400_000) + 1;
  return { startIso: toIsoDate(start), endIso: toIsoDate(end), totalDays };
}

export function quarterTotalWeeks(totalDays: number): number {
  return Math.min(MAX_WEEKS, Math.ceil(totalDays / 7));
}

/** Month-range label, e.g. Q1 → "Jan – Mar". */
export function quarterLabel(quarterNumber: number): string {
  const m = (quarterNumber - 1) * 3;
  return `${MONTHS[m]} – ${MONTHS[m + 2]}`;
}

export interface QuarterProgress {
  state: QuarterState;
  currentDay: number | null;
  currentWeek: number | null;
  totalDays: number;
  totalWeeks: number;
}

export function quarterProgress(year: number, quarterNumber: number, todayIso: string): QuarterProgress {
  const b = quarterBounds(year, quarterNumber);
  const totalWeeks = quarterTotalWeeks(b.totalDays);
  if (todayIso < b.startIso) {
    return { state: "UPCOMING", currentDay: null, currentWeek: null, totalDays: b.totalDays, totalWeeks };
  }
  if (todayIso > b.endIso) {
    return { state: "COMPLETED", currentDay: null, currentWeek: null, totalDays: b.totalDays, totalWeeks };
  }
  const dayIndex = Math.round(
    (parseIsoDate(todayIso).getTime() - parseIsoDate(b.startIso).getTime()) / 86_400_000,
  ); // 0-based
  return {
    state: "ACTIVE",
    currentDay: dayIndex + 1,
    currentWeek: Math.min(MAX_WEEKS, Math.floor(dayIndex / 7) + 1),
    totalDays: b.totalDays,
    totalWeeks,
  };
}

/** Which quarter number (1–4) contains a given date. */
export function quarterNumberOf(todayIso: string): number {
  const month = parseIsoDate(todayIso).getMonth(); // 0-based
  return Math.floor(month / 3) + 1;
}

/** Goal status, mirroring the backend QuarterMapper.status. */
export function computeGoalStatus(
  done: boolean,
  state: QuarterState,
  week: number,
  currentWeek: number | null,
): GoalStatus {
  if (done) return "DONE";
  if (state === "UPCOMING") return "UPCOMING";
  if (state === "COMPLETED") return "MISSED";
  // ACTIVE
  if (currentWeek != null && week === currentWeek) return "THIS_WEEK";
  return currentWeek != null && week > currentWeek ? "UPCOMING" : "MISSED";
}
